sync/src/chain.rs

```diff
@@ -2059,15 +2061,33 @@ impl ChainSync {
 						return None;
 					}
 
-					// Construct RLP
-					let mut packet = RlpStream::new_list(to_send.len());
-					for tx in &transactions {
-						if to_send.contains(&tx.transaction.hash()) {
-							packet.append(&tx.transaction);
-							// update stats
-							let id = io.peer_session_info(peer_id).and_then(|info| info.id);
-							stats.propagated(tx.transaction.hash(), id, block_number);
+
+					let mut to_send = to_send;
+					let mut packet;
+					loop {
+						// Construct RLP
+						packet = RlpStream::new_list(to_send.len());
```

Comment:
rather than halving the amount of propagated every time, I think we have a `RlpStream::new_list_unbounded` or similar that's used in snapshot creation. Then we can always fit as many as possible into the packet.